### DISCUSSION TITLE
[Bug] Suggestions of declared modules (via addExtraLib) are not showing up in suggestion window

### DISCUSSION THREAD

**Discussion by SlimShady0208** [microsoft/monaco-editor#3810] [Q&A]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
// Add additional d.ts files to the JavaScript language service and change.
// Also change the default compilation options.
// The sample below shows how a class Facts is declared and introduced
// to the system and how the compiler is told to use ES6 (target=2).

// validation settings
monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
	noSemanticValidation: false,
	noSyntaxValidation: false
});

// compiler options
monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
	target: monaco.languages.typescript.ScriptTarget.ES6,
	allowNonTsExtensions: true,
    noLib:true
});

// extra libraries
var libSource = [
	'declare module \'Module\' {',
	'   interface MyInterface {}',
    '   const fooBar:() => void',
	'}'
].join('\n');
//var libSource = [
//	'declare interface MyInterface {}',
//    'declare const fooBar:() => void',
//].join('\n');
var libUri = 'ts:filename/facts.d.ts';
monaco.languages.typescript.javascriptDefaults.addExtraLib(libSource, libUri);
// When resolving definitions and references, the editor will try to use created models.
// Creating a model for the library allows "peek definition/references" commands to work with the library.
monaco.editor.createModel(libSource, 'typescript', monaco.Uri.parse(libUri));

var jsCode = [
	'"use strict";',
	'',
	'class Chuck {',
	'    greet() {',
	'        return Facts.next();',
	'    }',
	'}'
].join('\n');

monaco.editor.create(document.getElementById('container'), {
	value: jsCode,
	language: 'javascript'
});
```

### Actual Behavior

- I am declaring a module `addExtraLib()`
- If I have fully typed the interface name, monaco-editor provides me a "Quick fix" to import the interface from "Module" (same for the `fooBar` const)
    - ![image](https://user-images.githubusercontent.com/949775/155725561-4de47fb4-1943-40d3-9ce7-c35ccb389a3a.png)

- But if I only type `const x:My`, the interface will not show up in the suggestion window
- If I change the extraLib, by removing the module declaration and declaring the interface/const directly (commented in the example) it will show up.
    - ![image](https://user-images.githubusercontent.com/949775/155726285-5346dc6d-c6a0-40f8-a4ee-30a25c4deafe.png)

### Expected Behavior

- I would expect the suggestions to show up in first case too or is this just impossible?
- I wouldn't even expect an automatic add of the import statement (this could be done by "Quick Fix"). But I would love to "lookup" my modules while typing. ;)

**Comment by SlimShady0208** [comment]
Any news on this one? Or did I miss to provide some more information? Or how could I help? ;)

**Comment by flokol120** [comment]
I've looked a bit deeper into this. Imagine the following setup within vscode:
```
│   package.json
│   tsconfig.json
│                           
├───src
│       index.ts
│       
└───types
        facts.d.ts
```
with `src/index.ts`:
```ts
class Chuck {
    greet() {
        const x: MyInte
        return "";
    }
}
```
and `types/facts.d.ts`:
```ts
declare module 'Module' {
    export interface MyInterface { }
    export const fooBar: () => void
}
```
While typing the import suggestion of `MyInterface` will work. By turning on the tsserver logging, you can also see this in the log:
```
...
Info 90   [08:57:24.820] response:
    {"seq":0,"type":"response","command":"completionEntryDetails","request_seq":12,"success":true,"body":[{"name":"MyInterface","kindModifiers":"export,declare","kind":"interface","displayParts":[{"text":"interface","kind":"keyword"},{"text":" ","kind":"space"},{"text":"MyInterface","kind":"interfaceName"}],"documentation":[],"tags":[],"codeActions":[{"description":"Add import from \"Module\"","changes":[{"fileName":"c:/local/fdo/Projects/module-test/src/index.ts","textChanges":[{"start":{"line":1,"offset":1},"end":{"line":1,"offset":1},"newText":"import { MyInterface } from \"Module\";\n"}]}]}],"source":[{"text":"Module","kind":"text"}],"sourceDisplay":[{"text":"Module","kind":"text"}]}]}
...
```
So this will work perfectly fine in vscode. Then I tried to replicate the same behavior in the monaco editor:
```ts
// validation settings
monaco.languages.typescript.typescriptDefaults.setDiagnosticsOptions({
	noSemanticValidation: false,
	noSyntaxValidation: false,
});

monaco.languages.typescript.typescriptDefaults.setModeConfiguration({
	definitions: true,
	completionItems: true,
	documentSymbols: true,
	codeActions: true,
	diagnostics: true,
	onTypeFormattingEdits: true,
});

// compiler options
monaco.languages.typescript.typescriptDefaults.setCompilerOptions({
	...monaco.languages.typescript.typescriptDefaults.getCompilerOptions(),
	target: monaco.languages.typescript.ScriptTarget.ES2016,
	module: monaco.languages.typescript.ModuleKind.CommonJS,
	moduleResolution: monaco.languages.typescript.ModuleResolutionKind.NodeJs,
	typeRoots: ["types"],
	rootDir: "./src"
});

// extra libraries
var libSource = [
	"declare module 'Module' {",
	'   export interface MyInterface {}',
    '   export const fooBar:() => void',
	'}'
].join('\n');

// won't work either!
// var libSource = [
// 	'export interface MyInterface {}',
//     'export const fooBar:() => void',
// ].join('\n');

var libUri = 'file:///types/facts.d.ts';
monaco.languages.typescript.typescriptDefaults.addExtraLib(libSource, libUri);

var tsCode = [
	'class Chuck {',
	'    greet() {',
	'        const x: MyInter',
	'        return "";',
	'    }',
	'}'
].join('\n');

var indexUri = 'file:///src/index.ts';
const model = monaco.editor.createModel(tsCode, 'typescript', monaco.Uri.parse(indexUri));

const editor = monaco.editor.create(document.getElementById('container'), {
	language: 'typescript',
	model,
});
```

This however won't work. I took some time debugging this and found out that, while `Module` is present in the `globals` map, the declarations within the module won't be in the `globals` map. Maybe that is the issue, as the `globals` map is later merged with the other found suggestions?

Interestingly enough, auto-importing works if I fully type out the name in the monaco editor, but not if I only partially type it :(

**Comment by xiaoxiyao** [comment]
It should be because of this https://github.com/microsoft/monaco-editor/issues/2582

**Comment by pieterjanv** [comment]
I was unable to reproduce the auto-import Quick Fix. However, I was able to implement my own auto-import for module declarations using a custom typescript worker and a completion item provider.

See https://github.com/pieterjanv/monaco-editor-module-auto-import-demo for a minimal example. The demo is at https://pieterjanv.github.io/monaco-editor-module-auto-import-demo.
